### PR TITLE
Adds SafeHTS library

### PR DESCRIPTION
**Pull request by stoqnkpL**
Adds `SafeHTS.sol` providing functionality allowing to execute safe operations on the HTS precompiled contract.
Adds `SafeOperations.sol` which showcases `SafeHTS.sol` usage. 

**Related issue(s)**:

Fixes https://github.com/hashgraph/hedera-smart-contracts/issues/43

**Notes for reviewer**:

**Checklist**

- [ ] Documented (Code comments, README, etc.)
- [ ] Tested (unit, integration, etc.)

### CODE REVIEW

**Review by shemnon — DISMISSED** [review]

File: safe-hts-precompile/SafeHTS.sol [hunk 1]
```diff
@@ -0,0 +1,458 @@
...
+
+    function safeCryptoTransfer(IHederaTokenService.TokenTransferList[] memory tokenTransfers) internal {
+        int32 responseCode;
+        (bool success, bytes memory result) = precompileAddress.call(
+            abi.encodeWithSelector(IHederaTokenService.cryptoTransfer.selector, tokenTransfers));
+        responseCode = success ? abi.decode(result, (int32)) : HederaResponseCodes.UNKNOWN;
+        require(responseCode == HederaResponseCodes.SUCCESS, "Safe crypto transfer failed!");
+    }
+
+    function safeMintToken(IHederaTokenService token, uint64 amount, bytes[] memory metadata) internal
```
Comment (shemnon) [inline safe-hts-precompile/SafeHTS.sol:22]: These should all go back to address. IHEderaTokenService is the interface for the precompile, not the type of the tokens.  These contracts are untyped.

**Review by shemnon — CHANGES_REQUESTED** [review]
All the address->IHederaTokenService changes that came in the force push need to be undone.

File: safe-hts-precompile/SafeHTS.sol (hunk 1 above)
Comment (MiroslavGatsanoga) [reply safe-hts-precompile/SafeHTS.sol]: Done.

Comment (shemnon) [reply safe-hts-precompile/SafeHTS.sol]: If the goal is to mimic OpenZeppelin's SafeERC20 then the name should stutter.  Example - https://github.com/code-423n4/2022-05-alchemix/blob/71abbe683dfd5c0686b7e594fb4f78a14b668d8b/contracts-full/adapters/fuse/FuseTokenAdapterV1.sol#L70

**Review by shemnon — DISMISSED** [review]

File: safe-hts-precompile/SafeHTS.sol [hunk 2]
```diff
@@ -0,0 +1,458 @@
...
+        (bool success, bytes memory result) = precompileAddress.call(
+            abi.encodeWithSelector(IHederaTokenService.cryptoTransfer.selector, tokenTransfers));
+        responseCode = success ? abi.decode(result, (int32)) : HederaResponseCodes.UNKNOWN;
+        require(responseCode == HederaResponseCodes.SUCCESS, "Safe crypto transfer failed!");
+    }
+
+    function safeMintToken(address token, uint64 amount, bytes[] memory metadata) internal
+    returns (uint64 newTotalSupply, int64[] memory serialNumbers) {
+        int32 responseCode;
+        (bool success, bytes memory result) = precompileAddress.delegatecall(
```
Comment (IvanKavaldzhiev) [inline safe-hts-precompile/SafeHTS.sol:25]: Do we deliberately use `delegatetcall` here?

File: safe-hts-precompile/SafeHTS.sol [hunk 3]
```diff
@@ -0,0 +1,458 @@
...
+        (responseCode, approved) = success ? abi.decode(result, (int32, address)) : (HederaResponseCodes.UNKNOWN, address(0));
+        require(responseCode == HederaResponseCodes.SUCCESS, "Safe get approved failed!");
+    }
+
+    function safeSetApprovalForAll(address token, address operator, bool approved) internal {
+        int32 responseCode;
+        (bool success, bytes memory result) = precompileAddress.call(
+            abi.encodeWithSelector(IHederaTokenService.setApprovalForAll.selector, token, operator, approved));
+        responseCode = success ? abi.decode(result, (int32)) : HederaResponseCodes.UNKNOWN;
+        require(responseCode == HederaResponseCodes.SUCCESS, "Safe set token approval failed!");
```
Comment (IvanKavaldzhiev) [inline safe-hts-precompile/SafeHTS.sol:219]: The error message could be changed to: "Safe set token approval for all failed!"

File: safe-hts-precompile/SafeHTS.sol [hunk 4]
```diff
@@ -0,0 +1,458 @@
...
+    }
+
+    function safeIsApprovedForAll(address token, address owner, address operator) internal
+    returns (bool approved)
+    {
+        int32 responseCode;
+        (bool success, bytes memory result) = precompileAddress.call(
+            abi.encodeWithSelector(IHederaTokenService.isApprovedForAll.selector, token, owner, operator));
+        (responseCode, approved) = success ? abi.decode(result, (int32, bool)) : (HederaResponseCodes.UNKNOWN, false);
+        require(responseCode == HederaResponseCodes.SUCCESS, "Safe query for if approved for all failed!");
```
Comment (IvanKavaldzhiev) [inline safe-hts-precompile/SafeHTS.sol:229]: `Safe is approved for all failed!`

File: safe-hts-precompile/SafeHTS.sol [hunk 5]
```diff
@@ -0,0 +1,458 @@
...
+    }
+
+    function safeIsFrozen(address token, address account) internal
+    returns (bool frozen)
+    {
+        int32 responseCode;
+        (bool success, bytes memory result) = precompileAddress.call(
+            abi.encodeWithSelector(IHederaTokenService.isFrozen.selector, token, account));
+        (responseCode, frozen) = success ? abi.decode(result, (int32, bool)) : (HederaResponseCodes.UNKNOWN, false);
+        require(responseCode == HederaResponseCodes.SUCCESS, "Safe query for if frozen failed!");
```
Comment (IvanKavaldzhiev) [inline safe-hts-precompile/SafeHTS.sol:239]: `Safe is frozen failed!`

File: safe-hts-precompile/SafeHTS.sol [hunk 6]
```diff
@@ -0,0 +1,458 @@
...
+    }
+
+    function safeIsKyc(address token, address account) internal
+    returns (bool kycGranted)
+    {
+        int32 responseCode;
+        (bool success, bytes memory result) = precompileAddress.call(
+            abi.encodeWithSelector(IHederaTokenService.isKyc.selector, token, account));
+        (responseCode, kycGranted) = success ? abi.decode(result, (int32, bool)) : (HederaResponseCodes.UNKNOWN, false);
+        require(responseCode == HederaResponseCodes.SUCCESS, "Safe query for if KYC is granted failed!");
```
Comment (IvanKavaldzhiev) [inline safe-hts-precompile/SafeHTS.sol:249]: `Safe is KYC granted failed!`

File: safe-hts-precompile/SafeHTS.sol [hunk 7]
```diff
@@ -0,0 +1,458 @@
...
+    returns (IHederaTokenService.FixedFee[] memory fixedFees, IHederaTokenService.FractionalFee[] memory fractionalFees, IHederaTokenService.RoyaltyFee[] memory royaltyFees)
+    {
+        int32 responseCode;
+        (bool success, bytes memory result) = precompileAddress.call(
+            abi.encodeWithSelector(IHederaTokenService.getTokenCustomFees.selector, token));
+        (responseCode, fixedFees, fractionalFees, royaltyFees) =
+        success
+        ? abi.decode(result, (int32, IHederaTokenService.FixedFee[], IHederaTokenService.FractionalFee[], IHederaTokenService.RoyaltyFee[]))
+        : (HederaResponseCodes.UNKNOWN, fixedFees, fractionalFees, royaltyFees);
+        require(responseCode == HederaResponseCodes.SUCCESS, "Safe query for token custom fees failed!");
```
Comment (IvanKavaldzhiev) [inline safe-hts-precompile/SafeHTS.sol:270]: `Safe get token custom fees failed!`

File: safe-hts-precompile/SafeHTS.sol [hunk 8]
```diff
@@ -0,0 +1,458 @@
...
+    }
+
+    function safeGetTokenDefaultFreezeStatus(address token) internal
+    returns (bool defaultFreezeStatus)
+    {
+        int32 responseCode;
+        (bool success, bytes memory result) = precompileAddress.call(
+            abi.encodeWithSelector(IHederaTokenService.getTokenDefaultFreezeStatus.selector, token));
+        (responseCode, defaultFreezeStatus) = success ? abi.decode(result, (int32, bool)) : (HederaResponseCodes.UNKNOWN, false);
+        require(responseCode == HederaResponseCodes.SUCCESS, "Safe query for token default freeze status failed!");
```
Comment (IvanKavaldzhiev) [inline safe-hts-precompile/SafeHTS.sol:280]: `Safe get token default freeze status failed!`

File: safe-hts-precompile/SafeHTS.sol [hunk 9]
```diff
@@ -0,0 +1,458 @@
...
+    }
+
+    function safeGetTokenDefaultKycStatus(address token) internal
+    returns (bool defaultKycStatus)
+    {
+        int32 responseCode;
+        (bool success, bytes memory result) = precompileAddress.call(
+            abi.encodeWithSelector(IHederaTokenService.getTokenDefaultKycStatus.selector, token));
+        (responseCode, defaultKycStatus) = success ? abi.decode(result, (int32, bool)) : (HederaResponseCodes.UNKNOWN, false);
+        require(responseCode == HederaResponseCodes.SUCCESS, "Safe query for token default KYC status failed!");
```
Comment (IvanKavaldzhiev) [inline safe-hts-precompile/SafeHTS.sol:290]: Safe get token default KYC status failed!

File: safe-hts-precompile/SafeHTS.sol [hunk 10]
```diff
@@ -0,0 +1,458 @@
...
+    }
+
+    function safeGetTokenExpiryInfo(address token) internal
+    returns (IHederaTokenService.Expiry memory expiry)
+    {
+        int32 responseCode;
+        (bool success, bytes memory result) = precompileAddress.call(
+            abi.encodeWithSelector(IHederaTokenService.getTokenExpiryInfo.selector, token));
+        (responseCode, expiry) = success ? abi.decode(result, (int32, IHederaTokenService.Expiry)) : (HederaResponseCodes.UNKNOWN, expiry);
+        require(responseCode == HederaResponseCodes.SUCCESS, "Safe query for token expiry info failed!");
```
Comment (IvanKavaldzhiev) [inline safe-hts-precompile/SafeHTS.sol:300]: `Safe get token expiry info failed!`

File: safe-hts-precompile/SafeHTS.sol [hunk 11]
```diff
@@ -0,0 +1,458 @@
...
+    }
+
+    function safeGetFungibleTokenInfo(address token) internal
+    returns (IHederaTokenService.FungibleTokenInfo memory fungibleTokenInfo)
+    {
+        int32 responseCode;
+        (bool success, bytes memory result) = precompileAddress.call(
+            abi.encodeWithSelector(IHederaTokenService.getFungibleTokenInfo.selector, token));
+        (responseCode, fungibleTokenInfo) = success ? abi.decode(result, (int32, IHederaTokenService.FungibleTokenInfo)) : (HederaResponseCodes.UNKNOWN, fungibleTokenInfo);
+        require(responseCode == HederaResponseCodes.SUCCESS, "Safe query for fungible token info failed!");
```
Comment (IvanKavaldzhiev) [inline safe-hts-precompile/SafeHTS.sol:310]: `Safe get fungible token info failed!`

File: safe-hts-precompile/SafeHTS.sol [hunk 12]
```diff
@@ -0,0 +1,458 @@
...
+    }
+
+    function safeGetTokenInfo(address token) internal
+    returns (IHederaTokenService.TokenInfo memory tokenInfo)
+    {
+        int32 responseCode;
+        (bool success, bytes memory result) = precompileAddress.call(
+            abi.encodeWithSelector(IHederaTokenService.getTokenInfo.selector, token));
+        (responseCode, tokenInfo) = success ? abi.decode(result, (int32, IHederaTokenService.TokenInfo)) : (HederaResponseCodes.UNKNOWN, tokenInfo);
+        require(responseCode == HederaResponseCodes.SUCCESS, "Safe query for token info failed!");
```
Comment (IvanKavaldzhiev) [inline safe-hts-precompile/SafeHTS.sol:320]: `Safe get token info failed!`

File: safe-hts-precompile/SafeHTS.sol [hunk 13]
```diff
@@ -0,0 +1,458 @@
...
+    }
+
+    function safeGetTokenKey(address token, uint keyType) internal
+    returns (IHederaTokenService.KeyValue memory key)
+    {
+        int32 responseCode;
+        (bool success, bytes memory result) = precompileAddress.call(
+            abi.encodeWithSelector(IHederaTokenService.getTokenKey.selector, token, keyType));
+        (responseCode, key) = success ? abi.decode(result, (int32, IHederaTokenService.KeyValue)) : (HederaResponseCodes.UNKNOWN, key);
+        require(responseCode == HederaResponseCodes.SUCCESS, "Safe query for token key failed!");
```
Comment (IvanKavaldzhiev) [inline safe-hts-precompile/SafeHTS.sol:330]: Safe get token key failed!

File: safe-hts-precompile/SafeHTS.sol [hunk 14]
```diff
@@ -0,0 +1,458 @@
...
+    }
+
+    function safeGetNonFungibleTokenInfo(address token, int64 serialNumber) internal
+    returns (IHederaTokenService.NonFungibleTokenInfo memory nonFungibleTokenInfo)
+    {
+        int32 responseCode;
+        (bool success, bytes memory result) = precompileAddress.call(
+            abi.encodeWithSelector(IHederaTokenService.getNonFungibleTokenInfo.selector, token, serialNumber));
+        (responseCode, nonFungibleTokenInfo) = success ? abi.decode(result, (int32, IHederaTokenService.NonFungibleTokenInfo)) : (HederaResponseCodes.UNKNOWN, nonFungibleTokenInfo);
+        require(responseCode == HederaResponseCodes.SUCCESS, "Safe query for NFT info failed!");
```
Comment (IvanKavaldzhiev) [inline safe-hts-precompile/SafeHTS.sol:340]: `Safe get non fungible token info failed!`

File: safe-hts-precompile/SafeHTS.sol [hunk 15]
```diff
@@ -0,0 +1,458 @@
...
+    }
+
+    function safeFreezeToken(address token, address account) internal {
+        int32 responseCode;
+        (bool success, bytes memory result) = precompileAddress.call(
+            abi.encodeWithSelector(IHederaTokenService.freezeToken.selector, token, account));
+        responseCode = success ? abi.decode(result, (int32)) : HederaResponseCodes.UNKNOWN;
+        require(responseCode == HederaResponseCodes.SUCCESS, "Safe freeze token operation failed!");
```
Comment (IvanKavaldzhiev) [inline safe-hts-precompile/SafeHTS.sol:348]: `Safe freeze token failed!`

File: safe-hts-precompile/SafeHTS.sol [hunk 16]
```diff
@@ -0,0 +1,458 @@
...
+    }
+
+    function safeUnfreezeToken(address token, address account) internal {
+        int32 responseCode;
+        (bool success, bytes memory result) = precompileAddress.call(
+            abi.encodeWithSelector(IHederaTokenService.unfreezeToken.selector, token, account));
+        responseCode = success ? abi.decode(result, (int32)) : HederaResponseCodes.UNKNOWN;
+        require(responseCode == HederaResponseCodes.SUCCESS, "Safe unfreeze token operation failed!");
```
Comment (IvanKavaldzhiev) [inline safe-hts-precompile/SafeHTS.sol:356]: `Safe unfreeze token failed!`

File: safe-hts-precompile/SafeHTS.sol [hunk 17]
```diff
@@ -0,0 +1,458 @@
...
+    }
+
+    function safeGrantTokenKyc(address token, address account) internal {
+        int32 responseCode;
+        (bool success, bytes memory result) = precompileAddress.call(
+            abi.encodeWithSelector(IHederaTokenService.grantTokenKyc.selector, token, account));
+        responseCode = success ? abi.decode(result, (int32)) : HederaResponseCodes.UNKNOWN;
+        require(responseCode == HederaResponseCodes.SUCCESS, "Safe grant token KYC operation failed!");
```
Comment (IvanKavaldzhiev) [inline safe-hts-precompile/SafeHTS.sol:364]: `Safe grant token KYC failed!`

File: safe-hts-precompile/SafeHTS.sol [hunk 18]
```diff
@@ -0,0 +1,458 @@
...
+    }
+
+    function safeRevokeTokenKyc(address token, address account) internal {
+        int32 responseCode;
+        (bool success, bytes memory result) = precompileAddress.call(
+            abi.encodeWithSelector(IHederaTokenService.revokeTokenKyc.selector, token, account));
+        responseCode = success ? abi.decode(result, (int32)) : HederaResponseCodes.UNKNOWN;
+        require(responseCode == HederaResponseCodes.SUCCESS, "Safe revoke token KYC operation failed!");
```
Comment (IvanKavaldzhiev) [inline safe-hts-precompile/SafeHTS.sol:372]: `Safe revoke token KYC failed!`

File: safe-hts-precompile/SafeHTS.sol [hunk 19]
```diff
@@ -0,0 +1,458 @@
...
+    }
+
+    function safePauseToken(address token) internal {
+        int32 responseCode;
+        (bool success, bytes memory result) = precompileAddress.call(
+            abi.encodeWithSelector(IHederaTokenService.pauseToken.selector, token));
+        responseCode = success ? abi.decode(result, (int32)) : HederaResponseCodes.UNKNOWN;
+        require(responseCode == HederaResponseCodes.SUCCESS, "Safe pause token operation failed!");
```
Comment (IvanKavaldzhiev) [inline safe-hts-precompile/SafeHTS.sol:380]: `Safe pause token failed!`

File: safe-hts-precompile/SafeHTS.sol [hunk 20]
```diff
@@ -0,0 +1,458 @@
...
+    }
+
+    function safeUnpauseToken(address token) internal {
+        int32 responseCode;
+        (bool success, bytes memory result) = precompileAddress.call(
+            abi.encodeWithSelector(IHederaTokenService.unpauseToken.selector, token));
+        responseCode = success ? abi.decode(result, (int32)) : HederaResponseCodes.UNKNOWN;
+        require(responseCode == HederaResponseCodes.SUCCESS, "Safe unpause token operation failed!");
```
Comment (IvanKavaldzhiev) [inline safe-hts-precompile/SafeHTS.sol:388]: `Safe unpause token failed!`

File: safe-hts-precompile/SafeHTS.sol [hunk 21]
```diff
@@ -0,0 +1,458 @@
...
+    }
+
+    function safeWipeTokenAccount(address token, address account, uint32 amount) internal {
+        int32 responseCode;
+        (bool success, bytes memory result) = precompileAddress.call(
+            abi.encodeWithSelector(IHederaTokenService.wipeTokenAccount.selector, token, account, amount));
+        responseCode = success ? abi.decode(result, (int32)) : HederaResponseCodes.UNKNOWN;
+        require(responseCode == HederaResponseCodes.SUCCESS, "Safe wipe token account operation failed!");
```
Comment (IvanKavaldzhiev) [inline safe-hts-precompile/SafeHTS.sol:396]: `Safe wipe token account failed!`

File: safe-hts-precompile/SafeHTS.sol [hunk 22]
```diff
@@ -0,0 +1,458 @@
...
+    }
+
+    function safeWipeTokenAccountNFT(address token, address account, int64[] memory serialNumbers) internal {
+        int32 responseCode;
+        (bool success, bytes memory result) = precompileAddress.call(
+            abi.encodeWithSelector(IHederaTokenService.wipeTokenAccountNFT.selector, token, account, serialNumbers));
+        responseCode = success ? abi.decode(result, (int32)) : HederaResponseCodes.UNKNOWN;
+        require(responseCode == HederaResponseCodes.SUCCESS, "Safe wipe token account NFT operation failed!");
```
Comment (IvanKavaldzhiev) [inline safe-hts-precompile/SafeHTS.sol:404]: `Safe wipe token account NFT failed!`

File: safe-hts-precompile/SafeHTS.sol [hunk 23]
```diff
@@ -0,0 +1,458 @@
...
+    }
+
+    function safeUpdateTokenInfo(address token, IHederaTokenService.HederaToken memory tokenInfo) nonEmptyExpiry(tokenInfo) internal {
+        int32 responseCode;
+        (bool success, bytes memory result) = precompileAddress.call(
+            abi.encodeWithSelector(IHederaTokenService.updateTokenInfo.selector, token, tokenInfo));
+        responseCode = success ? abi.decode(result, (int32)) : HederaResponseCodes.UNKNOWN;
+        require(responseCode == HederaResponseCodes.SUCCESS, "Safe update token info operation failed!");
```
Comment (IvanKavaldzhiev) [inline safe-hts-precompile/SafeHTS.sol:412]: `Safe update token info failed!`

File: safe-hts-precompile/SafeHTS.sol [hunk 24]
```diff
@@ -0,0 +1,458 @@
...
+    }
+
+    function safeUpdateTokenExpiryInfo(address token, IHederaTokenService.Expiry memory expiryInfo) internal {
+        int32 responseCode;
+        (bool success, bytes memory result) = precompileAddress.call(
+            abi.encodeWithSelector(IHederaTokenService.updateTokenExpiryInfo.selector, token, expiryInfo));
+        responseCode = success ? abi.decode(result, (int32)) : HederaResponseCodes.UNKNOWN;
+        require(responseCode == HederaResponseCodes.SUCCESS, "Safe update token expiry info operation failed!");
```
Comment (IvanKavaldzhiev) [inline safe-hts-precompile/SafeHTS.sol:420]: `Safe update token expiry info failed!`

File: safe-hts-precompile/SafeHTS.sol [hunk 25]
```diff
@@ -0,0 +1,458 @@
...
+    }
+
+    function safeUpdateTokenKeys(address token, IHederaTokenService.TokenKey[] memory keys) internal {
+        int32 responseCode;
+        (bool success, bytes memory result) = precompileAddress.call(
+            abi.encodeWithSelector(IHederaTokenService.updateTokenKeys.selector, token, keys));
+        responseCode = success ? abi.decode(result, (int32)) : HederaResponseCodes.UNKNOWN;
+        require(responseCode == HederaResponseCodes.SUCCESS, "Safe update token keys operation failed!");
```
Comment (IvanKavaldzhiev) [inline safe-hts-precompile/SafeHTS.sol:428]: `Safe update token keys failed!`

File: safe-hts-precompile/SafeHTS.sol [hunk 26]
```diff
@@ -0,0 +1,458 @@
...
+    }
+
+    function safeIsToken(address token) internal
+    returns (bool isToken)
+    {
+        int32 responseCode;
+        (bool success, bytes memory result) = precompileAddress.call(
+            abi.encodeWithSelector(IHederaTokenService.isToken.selector, token));
+        (responseCode, isToken) = success ? abi.decode(result, (int32, bool)) : (HederaResponseCodes.UNKNOWN, false);
+        require(responseCode == HederaResponseCodes.SUCCESS, "Safe query for if valid token at address failed!");
```
Comment (IvanKavaldzhiev) [inline safe-hts-precompile/SafeHTS.sol:438]: `Safe is token failed!`

File: safe-hts-precompile/SafeHTS.sol [hunk 27]
```diff
@@ -0,0 +1,458 @@
...
+    }
+
+    function safeGetTokenType(address token) internal
+    returns (int32 tokenType)
+    {
+        int32 responseCode;
+        (bool success, bytes memory result) = precompileAddress.call(
+            abi.encodeWithSelector(IHederaTokenService.getTokenType.selector, token));
+        (responseCode, tokenType) = success ? abi.decode(result, (int32, int32)) : (HederaResponseCodes.UNKNOWN, int32(0));
+        require(responseCode == HederaResponseCodes.SUCCESS, "Safe query for token type failed!");
```
Comment (IvanKavaldzhiev) [inline safe-hts-precompile/SafeHTS.sol:448]: Safe get token type failed!

File: safe-hts-precompile/SafeOperations.sol [hunk 28]
```diff
@@ -0,0 +1,330 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity >=0.5.0 <0.9.0;
+pragma experimental ABIEncoderV2;
+
+import "./SafeHTS.sol";
+
+contract SafeOperations {
+
+    function safeTokenAssociate(address sender, address tokenAddress) external {
```
Comment (IvanKavaldzhiev) [inline safe-hts-precompile/SafeOperations.sol:9]: You can keep the same names for all methods as they are in the SafeHTS library, e.g. `safeAssociateToken`

File: safe-hts-precompile/SafeHTS.sol (hunk 2 above)
Comment (MiroslavGatsanoga) [reply safe-hts-precompile/SafeHTS.sol:25]: It was not added in the commits made by me, but looking at the [hedera-services PR](https://github.com/hashgraph/hedera-services/pull/2866) that precedes this PR it seems to be fixed there.

I don't see an obvious reason to use `delegatecall` here so I will change it to `call`.

**Review by IvanKavaldzhiev — DISMISSED** [review]
LGTM

File: safe-hts-precompile/SafeHTS.sol [hunk 29]
```diff
@@ -435,7 +435,7 @@ library SafeHTS {
         (bool success, bytes memory result) = precompileAddress.call(
             abi.encodeWithSelector(IHederaTokenService.isToken.selector, token));
         (responseCode, isToken) = success ? abi.decode(result, (int32, bool)) : (HederaResponseCodes.UNKNOWN, false);
-        require(responseCode == HederaResponseCodes.SUCCESS, "Safe query for if valid token at address failed!");
+        require(responseCode == HederaResponseCodes.SUCCESS, "Safe is token failed!");
```
Comment (shemnon) [inline safe-hts-precompile/SafeHTS.sol:438]: This reads weird.  While all the other variants split out the words I think we should consider keeping this camelCase - `Safe isToken failed!` or `Safe 'is token' failed`

**Review by shemnon — DISMISSED** [review]

**Review by Nana-EC — APPROVED** [review]

File: safe-hts-precompile/SafeHTS.sol (hunk 29 above)
Comment (MiroslavGatsanoga) [reply safe-hts-precompile/SafeHTS.sol]: Updated.

**Review by IvanKavaldzhiev — APPROVED** [review]
LGTM

**Review by shemnon — APPROVED** [review]

**Pull request merged**